Frozen truncate

jordanthomas opened the pull request:
### Summary

Return from truncate is inconsistent. When the string is not long enough to be truncated an unfrozen string is returned. When it is long enough to be truncated a new frozen string is returned.

Fixes https://github.com/rails/rails/issues/36106.

rafaelfranca commented on activesupport/lib/active_support/core_ext/string/filters.rb:
```diff
@@ -64,7 +64,7 @@ def remove!(*patterns)
   #   # => "And they f... (continued)"
   def truncate(truncate_at, options = {})
-    return dup unless length > truncate_at
+    return dup.freeze unless length > truncate_at
```
I think it should be the opposite, we should always return a non-frozen string. In the same way as `gsub` and `strip` don't. I believe it only return frozen strings because of the `frozen_string_literal` pragm in this file and because we are using interpolation before returning.

rafaelfranca replied:
```
>> 'foo'.freeze.gsub('a', '').frozen?
=> false
>> 'foo'.freeze.gsub('o', '').frozen?
=> false
```

jordanthomas replied:
Ah right on, I think that's even more predictable. I wrongly interpreted that the `frozen_string_literal` thing was trying to ensure returns were frozen.

I'll swap it.

jordanthomas committed:
Add test asserting truncate returns unfrozen string

jordanthomas committed:
Ensure strings returned from truncate are not frozen

This fixes an issue where strings too short to be truncated were
returned unfrozen, where as long-enough strings were returned
frozen. Now retuned strings will not be frozen whether or not
the string returned was shortened.

jordanthomas committed:
Update changelog w/ new truncate behavior description

rafaelfranca commented:
Thank you!
